[RW-8041][risk=low]Added the ability to use the custom request access url from the regis…

evrii opened the pull request:
Description: Added the ability to use the custom request access url from the registration page when a registrant attempts to register for an institute that requires only specific email addresses.


https://user-images.githubusercontent.com/24514630/159502948-0d6cdff7-b330-4ba3-8e45-9d01dbadd96f.mov

https://user-images.githubusercontent.com/24514630/159574644-f0a9f05b-eb75-428e-af7e-937c7e9950be.mov



<!--
Replace this template with your PR description.
Please remember to keep in mind the security levels outlined in
[CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md) and to
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data
* **severe**: Severe chance of potential impact to, or exposure of patient data

Please also:

* Get thumbs from reviewer(s)
* Verify all tests go green, including CI tests
-->


---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

jmthibault79 commented on api/src/main/resources/workbench-api.yaml:
```diff
@@ -6913,6 +6913,9 @@ definitions:
          or restricted to a defined set of specific researchers'
+      requestAccessUrl:
+        type: string
+        description: Request access url for users to use during registration.
```
please note that this is optional

evrii replied:
I updated the description. Thanks for pointing this out.

jmthibault79 commented on ui/src/app/utils/institutions.tsx:
```diff
@@ -42,16 +43,22 @@ export async function checkInstitutionalEmail(
 }
 
-const EmailAddressMismatchErrorMessage = () => {
+const EmailAddressMismatchErrorMessage = ({ requestAccessUrl }) => {
```
this is a nice extension of the existing version.

jmthibault79 reviewed (commented):
Looks good so far.  There is one piece that remains however: we need to also use this link for the "Request Access" to Controlled Tier button on the Data Access Requirements page:
<img width="559" alt="Screen Shot 2022-03-22 at 1 58 25 PM" src="https://user-images.githubusercontent.com/2701406/159545566-76f43f0c-8f8d-497d-8e5c-3da1db23c933.png">

evrii commented:
> Looks good so far. There is one piece that remains however: we need to also use this link for the "Request Access" to Controlled Tier button on the Data Access Requirements page: <img alt="Screen Shot 2022-03-22 at 1 58 25 PM" width="559" src="https://user-images.githubusercontent.com/2701406/159545566-76f43f0c-8f8d-497d-8e5c-3da1db23c933.png">

Thanks for pointing this out. I added the functionality to work on this page as well. I updated the description of the PR to include a recording of how this screen works as well.

jmthibault79 commented on api/src/main/resources/workbench-api.yaml:
```diff
@@ -6878,7 +6878,7 @@ definitions:
       requestAccessUrl:
         type: string
-        description: Request access url for users to use during registration.
+        description: Optional url for institution's handling their own access requests
```
sometimes apostrophes here can cause problems.  did you notice anything running locally?

jmthibault79 replied:
(I did not)

evrii replied:
Good point. I do not recall seeing any issues. Agreed though, apostrophes can be tricky.

jmthibault79 commented on ui/src/app/components/support.tsx:
```diff
@@ -10,8 +12,21 @@ export const SupportMailto = ({ label = SUPPORT_EMAIL, style = {} }) => (
-export const SupportButton = ({ label = SUPPORT_EMAIL, style = {} }) => (
-  <Button style={style} onClick={() => window.open(`mailto:${SUPPORT_EMAIL}`)}>
+const handleClickSupportButton = (url) => () => {
+  if (url) {
```
this logic is now repeated - would it make sense to reuse some code here?

evrii replied:
Yeah, I wonder if this functionality should live in the urls utils file.

jmthibault79 commented on ui/src/app/pages/homepage/data-access-requirements.tsx:
```diff
@@ -952,7 +955,10 @@ const ControlledTierCard = (props: {
             </div>
             <div style={styles.requestAccess}>
-              <SupportButton label='Request Access' />
+              <SupportButton
```
I like how concise this is.  But I'm not sure whether this use case should live inside SupportButton - it seems to me like it's pretty specific to D-A-R.  What do you think?

evrii replied:
Good point. I did notice that SupportButton is only used in D-A-R. I am good with moving it.

jmthibault79 reviewed (approved):
Tested locally.  Works great.

Very nice recordings.

Your DB workflow looks nicer than mine - I'd love to learn more about that sometime.

evrii commented on ui/src/app/utils/urls.ts:
```diff
@@ -9,3 +9,23 @@ export function canonicalizeUrl(url: string): string {
+  }
+}
+
+export function getCustomOrDefaultUrl(customUrl: string, defaultUrl: string) {
```
I am unsure of the name of this function. I am open to suggestions if anyone can think of anything better.

NehaBroad reviewed (approved):
Test for ByPass/Expiring Soon and Expired

NehaBroad reviewed (approved):
Tested 👍
